core/trino-main/src/main/java/io/trino/sql/planner/ConnectorExpressionTranslator.java

```diff
@@ -297,6 +335,32 @@ public SqlToConnectorExpressionTranslator(Session session, Map<NodeRef<Expressio
             return Optional.of(new FieldDereference(typeOf(node), translatedBase.get(), (int) (((LongLiteral) node.getIndex()).getValue() - 1)));
         }
 
+        @Override
+        protected Optional<ConnectorExpression> visitInPredicate(InPredicate node, Void context)
+        {
+            InListExpression valueList = (InListExpression) node.getValueList();
+            Optional<ConnectorExpression> valueExpression = process(node.getValue());
+
+            if (valueExpression.isEmpty()) {
+                return Optional.empty();
+            }
+
+            ImmutableList.Builder<ConnectorExpression> values = ImmutableList.builder();
+
+            for (Expression value : valueList.getValues()) {
+                Optional<ConnectorExpression> processedValue = process(value);
+
+                if (processedValue.isEmpty()) {
+                    return Optional.empty();
+                }
+
+                values.add(processedValue.get());
+            }
+
+            ConnectorExpression inListExpression = new Call(typeOf(node.getValueList()), StandardFunctions.IN_LIST_FUNCTION_NAME, values.build());
+            return Optional.of(new Call(typeOf(node), StandardFunctions.IN_PREDICATE_FUNCTION_NAME, List.of(valueExpression.get(), inListExpression)));
```

Comment:
I think we need to think about efficiency of the representation, since IN list can be long and can be processed multiple times. This is the same problem as with large Domains, and with same performance challenges. See 82e8f93ab3149e56b2daed4b5102423a953e9a9f and cc @raunaqmorarka 

Now, I am aware that `InListExpression` doesn't have an efficient representation, but it can be evolved to have one. With `ConnectorExpression`s, such an evolution would be an SPI breaking change, so I'd suggest we consider performance in the first version already.

cc @martint

Reply:
@martint any thoughts here?